src/tools/rbd_mirror/image_replayer/BootstrapRequest.cc

```diff
@@ -222,6 +224,11 @@ void BootstrapRequest<I>::handle_prepare_remote_image(int r) {
     // if the local image is linked to the remote image, we ignore that
     // the remote image is not primary so that we can replay demotion
   }
+  if (m_skip_sync) {
+      dout(10) << "skipping image sync" << dendl;
+      finish(-ECANCELED);
```

Comment:
> and call `cancel()` before `send()`.

Looking deeper, if we do this, I would argue that we don't need to change `BootstrapRequest` at all:
- as soon as `m_bootstrap_request` pointer is assigned in `ImageReplayer::bootstrap()`, we are in a common "image replayer is interrupted while bootstrap is in flight" scenario.  Here `ImageReplayer::stop()` would call `cancel()`, and whether that happens a millisecond later or two seconds later is irrelevant.
- by calling `cancel()` before `send()` in case `m_stop_requested` is set we are taking care of a race window where `ImageReplayer::stop()` can be called after `m_state` is set to `STATE_STARTING` in `ImageReplayer::start()` but before `m_bootstrap_request` pointer is assigned in `ImageReplayer::bootstrap()`.  Here `ImageReplayer::stop()` would not call `cancel()`, so we are compensating.

Since `BootstrapRequest::cancel()` propagates to `ImageSync::cancel()` which in turn propagates to `SnapshotCopyRequest::cancel()` and `ImageCopyRequest::cancel()`, we should be OK with `BootstrapRequest` as is (assuming all this cancellation code is working -- if it doesn't, that is out of scope of this PR).  Going out of the way to bail out of `BootstrapRequest` slightly sooner isn't necessary.

Reply:
Done.